src/remoteConnector.ts

```diff
@@ -771,62 +791,64 @@ export default class RemoteConnector extends Disposable {
-				await vscode.commands.executeCommand('__gitpod.getGitpodRemoteLogsUri');
-				return true;
-			} catch {
-				return false;
-			}
-		};
+	private startHeartBeat(accessToken: string, connectionInfo: SSHConnectionParams) {
+		if (this.heartbeatManager) {
+			return;
+		}
 
-		const parseSSHDest = (sshDestStr: string): { user: string; hostName: string } | string => {
-			let decoded;
-			try {
-				decoded = JSON.parse(Buffer.from(sshDestStr, 'hex').toString('utf8'));
-			} catch {
-				// no-op
-			}
-			return decoded && typeof decoded.hostName === 'string' ? decoded : sshDestStr;
-		};
+		this.heartbeatManager = new HeartbeatManager(connectionInfo.gitpodHost, connectionInfo.workspaceId, connectionInfo.instanceId, accessToken, this.logger, this.telemetry);
+	}
 
+	private async onGitpodRemoteConnection() {
 		const remoteUri = vscode.workspace.workspaceFile || vscode.workspace.workspaceFolders?.[0].uri;
-		if (vscode.env.remoteName === 'ssh-remote' && this.context.extension.extensionKind === vscode.ExtensionKind.UI && remoteUri) {
-			const [, sshDestStr] = remoteUri.authority.split('+');
-			const sshDest = parseSSHDest(sshDestStr);
-
-			const connectionSuccessful = await isRemoteExtensionHostRunning();
-			const connectionInfo = this.context.globalState.get<SSHConnectionParams>(`${RemoteConnector.SSH_DEST_KEY}${sshDestStr}`);
-			if (connectionInfo) {
-				sshDest;
-				// const usingSSHGateway = typeof sshDest !== 'string';
-				// const kind = usingSSHGateway ? 'gateway' : 'local-app';
-				// if (connectionSuccessful) {
-				// 	this.telemetry.sendTelemetryEvent('vscode_desktop_ssh', {
-				// 		kind,
-				// 		status: 'connected',
-				// 		instanceId: connectionInfo.instanceId,
-				// 		workspaceId: connectionInfo.workspaceId,
-				// 		gitpodHost: connectionInfo.gitpodHost
-				// 	});
-				// } else {
-				// 	this.telemetry.sendTelemetryEvent('vscode_desktop_ssh', {
-				// 		kind,
-				// 		status: 'failed',
-				// 		reason: 'remote-ssh extension: connection failed',
-				// 		instanceId: connectionInfo.instanceId,
-				// 		workspaceId: connectionInfo.workspaceId,
-				// 		gitpodHost: connectionInfo.gitpodHost
-				// 	});
-				// }
-				await this.context.globalState.update(`${RemoteConnector.SSH_DEST_KEY}${sshDestStr}`, undefined);
-			}
+		if (!remoteUri) {
+			return;
+		}
+
+		const [, sshDestStr] = remoteUri.authority.split('+');
+		const connectionInfo = this.context.globalState.get<SSHConnectionParams & { isFirstConnection: boolean }>(`${RemoteConnector.SSH_DEST_KEY}${sshDestStr}`);
+		if (!connectionInfo) {
+			return;
+		}
 
-			return connectionSuccessful;
+		await this.context.globalState.update(`${RemoteConnector.SSH_DEST_KEY}${sshDestStr}`, { ...connectionInfo, isFirstConnection: false });
+
+		// Check for gitpod remote extension version to avoid sending heartbeat in both extensions at the same time
+		const isGitpodRemoteHeartbeatCancelled = await cancelGitpodRemoteHeartbeat();
```

Comment:
I'm not sure about it. We see that remote-ssh-extension fails to install right now very often. If we do it then heartbeat is missing.